[#417] Playwright Recorder + Replayer

AnasSaleh-98 opened the pull request:
## Description
This feature is to allow users to record and replay browser workflows directly within Semoss. The workflow is initiated by recording a flow, generating a json file with all the steps done, and retrieving those files for replay whenever the user needs to integrate with any external system. These reactors are utilized and getting called by a react app that is going to be exposed as an MCP tool in playground.

## Changes Made
<!-- List the key changes made in this PR -->
1. Added a new package called **perna.reactor.playwright** that contains all the reactors and files needed.
2. Added a utility method **getMap** in perna.util.Utility.java as it is used by multiple reactors in playwright (and duplicated in other reactors as well)

## How to Test

1. Call Session ( ) to open a new session and use the sessionId returned in future pixel calls.
2. Call Step reactor with the session id returned to test different actions user can make (click, navigate, type, etc.)
3. Call SaveAll reactor to save the steps. (Note that currently the path is relative to project app folder, and so when testing this from terminal it will throw an error since the project name is null. **For testing purposes, you can change the folder path to any local folder to see the json file by editing the path in this reactor :ReplayFromFileReactor: initRecordingsDir**)

AnasSaleh-98 committed:
[#417] feat: added playwright reactors

AnasSaleh-98 committed:
[#417] feat: updated playwright reactors

RawanAbdelkhalek committed:
Merge remote-tracking branch 'origin/dev' into playwright-reactors

pasupathimuniyappan committed:
[1235] - feat - Add JSON structured support for Anthropic and Bedrock Clients. (#1259)

* [1235] - feat - Added the support for JSON structured responses for Anthropic and Bedrock Clients.

* [1235] - feat - standardized the code.

* [1235] - feat - included the script to standardize the response of bedrock and anthropic clients for JSON structured.

pasupathimuniyappan committed:
[1239] - feat - Added the new message format support for AzureOpenAiChatCompletion and AzureOpenAIResponses through OpenAiClientV2 class. (#1267)

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

ryanweiler92 committed:
feat: update model urls (#1271)

themaherkhalil committed:
fix: modification for processing java to python string (#1270)

tlokeshrao committed:
Reduce Clauses in Bool OpenSearch (#1268)

* fix: reduce clauses in bool

* fix: terms

* fix: comparator

* fix: pixeldatatype

kunal0137 committed:
fix: casting to number to get long value (#1276)

kunal0137 committed:
fix-casting-modellogs-2 (#1277)

themaherkhalil committed:
feat: exposing ability to generate mcp for just a single tool that gets appended (instead of replacing) the current mcp json (#1280)

* feat: exposing ability to generate mcp for just a single tool that gets appended (instead of replacing) the current mcp json

* fix: proper error messages for the expected file

Varaham committed:
feat: exposing admin reactor to view RDF contents (#1274)

* added code for masking Sensitive data in RDF Map prop file

* feat: added code for masking RDF Map sensitive  data

* chore: moving methods to DIHelper directly instead of Utility

* chore: renaming reactor and adding description

---------

Co-authored-by: Varaham <katchabi50@gmail.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

AnasSaleh-98 committed:
[#417] feat: updated playwright reactors

AnasSaleh-98 committed:
Merge branch 'playwright-reactors' of https://github.com/SEMOSS/Semoss into playwright-reactors

pasupathimuniyappan committed:
[1266] - feat - Support beta headers for the Python Anthropic Client (#1281)

* [1266] - feat - Added the support to Anthropic client for the beta headers to support the large documents.

* [1266] - feat - Updated the code.

* feat: handle betas with streaming legacy messages

---------

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

ryanweiler92 committed:
Don't require beta headers param for anthropic models (#1282)

* feat: don't require beta headers param for anthropic models

* feat: update use_beta_header typing

AnasSaleh-98 committed:
[#417] feat: added variable record with isPassword field

pmagudapathy committed:
feat: adding log4j2.xml and new appenders to log into audit database (#1208)

* feat: creation of the audit logs db on startup to hold both the server logs and the engine method logs performed through PipelineInvocationHandler

* feat: enabling log4j2.xml at the engine level within the assets folder to separate out logs. Example could be

<?xml version="1.0" encoding="UTF-8"?>
<Configuration xmlns:xi="http://www.w3.org/2001/XInclude" xmlns="https://logging.apache.org/xml/ns"
	status="INFO"
	packages="prerna.logging"
	monitorInterval="30">
	<Appenders>
		<!-- Console Appender -->
		<Console name="Console" target="SYSTEM_OUT">
			<PatternLayout>
				<!-- <Pattern>[%-5level] %d{yyyy-MM-dd HH:mm:ss} %c{1.}:%L [user=%X{userId}] [session=%X{sessionId}] [ip=%X{clientIP}] %m%n</Pattern> -->
				<Pattern>CUSTOM_ENGINE [%-5level] %d{yyyy-MM-dd HH:mm:ss} %c{1.}:%L [user=%X{userId}] %m%n</Pattern>
			</PatternLayout>
		</Console>

		<!-- Database Appender -->
        <ServerLogsJDBCAppender name="ServerLogsJDBCAppender" batchSize="1" ignoreExceptions="false">
        	<PatternLayout>
				<Pattern>CUSTOM_ENGINE [%-5level] %d{yyyy-MM-dd HH:mm:ss} %c{1.}:%L [user=%X{userId}] [session=%X{sessionId}] [ip=%X{clientIP}] [reqId=%X{requestId}] %m%n</Pattern>
			</PatternLayout>
        </ServerLogsJDBCAppender>
        <AuditLogsJDBCAppender name="AuditLogsJDBCAppender" batchSize="1" ignoreExceptions="false" engineId="11629219-1fbf-4b52-af2a-04b509b2f111">
       		<PatternLayout>
				<Pattern>CUSTOM_ENGINE [%-5level] %d{yyyy-MM-dd HH:mm:ss} %c{1.}:%L [user=%X{userId}] [session=%X{sessionId}] [ip=%X{clientIP}] [reqId=%X{requestId}] %m%n</Pattern>
			</PatternLayout>
        </AuditLogsJDBCAppender>
	</Appenders>
	<Loggers>
		<AsyncLogger name="EngineLogger" level="INFO" additivity="false" includeLocation="false">
			<AppenderRef ref="AuditLogsJDBCAppender" />
			<AppenderRef ref="Console" />
			<AppenderRef ref="ServerLogsJDBCAppender" />
		</AsyncLogger>
		<AsyncRoot level="WARN" includeLocation="false">
			<AppenderRef ref="Console" />
		</AsyncRoot>
	</Loggers>
</Configuration>

---------

Co-authored-by: US\pkapaleeswaran <pkapaleeswaran@deloitte.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>
Co-authored-by: Magudapathy <pmagudapathy@deloitte.com>

themaherkhalil committed:
feat: exposing admin only checks on guardrails (#1286)

* feat: exposing admin only checks on guardrails

* chore: adding admin values for guardrails in rdf_map

* chore: adding secrets for guardrails

AnasSaleh-98 committed:
[#417] feat: fixed playwright variables not retrieved when text is empty

themaherkhalil committed:
feat: mcp notebook will now replace the python function in the driver python file if there is an existing mcp tool that has that cell id in its meta (#1288)

Subhadeepghosh1 committed:
Fix upload issue with proper cleanup from DIHelper (#1272)

* fix: fix database file lock issue

* fix: removing engine from DIHelper

* fix: applying fix on upload of projects, removing db logic

---------

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

kunal0137 committed:
fix-remove-microservice-url in RDF (#1291)

themaherkhalil committed:
fix: switch from athena simba jar to aws jar (#1293)

themaherkhalil committed:
chore: switching code creation to text block instead of concatenations (#1294)

themaherkhalil committed:
fix: adding plugin to remove log4j from athena fat jar (#1297)

themaherkhalil committed:
fix: eclipse runs with a specific class loader, need changes to ensure it works for development (#1298)

ryanweiler92 committed:
Upgrading Python Packages (#1295)

* feat: upgrading faiss-cpu && numpy

* feat: upgrading openai, anthropic, googlegenai, pydantic adding langextract

---------

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
chore: removing deprecated plugin lookup attribute (#1299)

themaherkhalil committed:
chore: remove aws v1 sdk (#1300)

* chore: removing old aws jars and s3 custom reactors - use storage reactors instead

* chore: ignoring dependency-reduced-pom.xml

themaherkhalil committed:
fix: delete classes associated with aws v1 code that was removed (#1302)

tlokeshrao committed:
fix throw query errors (#1304)

Subhadeepghosh1 committed:
feat: adding logic for custom groups, formatting security util code, try-catch-finally to try-with-resource (#1220)


---------

Co-authored-by: Anurag Jain <anurag91jain@gmail.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
chore: switching to try-with-resource (#1305)

kaviiKS committed:
feat: google ocr,  aws textract, pom dependency updates (#1189)

* feat: exposing new OCR routines that can be integrated with vector databases

* feat: removing com.microsoft.azure jars (legacy v1) for com.azure jars

* chore: removing slf4j logger factory for log4j log manager

* feat: deprecating CreateRestFunctionEngineReactor for CreateFunctionEngineReactor

* chore: with introduction of new jars and dependencies growing, modified plugin to require 100% dependency convergence. generate report by running:
 
mvn project-info-reports:dependency-convergence

---------

Co-authored-by: kavks <Kavks@deloitte.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
chore: exposing mvn site to get dependency list and dependency coverage (#1306)

themaherkhalil committed:
chore: updating jars, switching scheduler to hikari from c3p0 (#1307)

pasupathimuniyappan committed:
[1232] - feat - Add Bedrock support through Anthropic's SDK (#1285)

* [1232] - feat - added the bedrock support through anthropic's sdk.

* [1232] - feat - Standardize the code.

* feat: instantiate anthropic  bedrock client in anthropic class instead of google

* feat: removing ref to bedrock client in google client class

---------

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

kunal0137 committed:
fix: installing gliner (#1311)

kunal0137 committed:
Mesage table structure rework message (#1195)

* fix: add transaction id column

* fix: message table structure for new messages

* fix: compilation

* fix: update old rooms

* fix: other entry points

* fix: theoretically good

* fix: updation

* fix: extra uuid

* fix: other things

* fix: using message id from AbstractMessage

* fix: imports

* fix: migrate ids and overload doRecordMessage

---------

Co-authored-by: Tejas Lokeshrao <tlokeshrao@deloitte.com>
Co-authored-by: rithvik-doshi <81876806+rithvik-doshi@users.noreply.github.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
feat: switching to new mcp_driver notebook and mcp_driver.py from smss_driver and smss_driver.py (#1309)

themaherkhalil committed:
feat: switching from UUID to GUID.v7() (#1314)

ryanweiler92 committed:
OpenAI transcription function engine (#1284)

* feat: openai transcription function engine

* feat: wip

* feat: OpenAI Transcribe Function Engine

kunal0137 committed:
Cot playground (#1290)

* fix: create step class

* feat: add cotroomreactor outline

* chore: separating playground items

* feat: adding initial cot room

* feat: update updateroomoptionsreactor with other version

* fix: multiple vec db

* fix: correct keyargs

* feat: update room in memory with options

* feat: tools in room options

* fix: pushing cot changes

* fix: check type on python to handle list strs

* fix: proper json return for complex data type

* feat: add step class and adjust cot templates accordingly

* fix: small comment for util

* fix: partial refactor of step class

* feat: finish refactor of step class

* fix: move success criteria to seperate class and consolidate builders

* feat: add prompts to utils

* reverting prompt structure

* fix: triple quote strings incase they have quotes

* feat: create triage reactor

* feat: modify output configuration

* fix: import file

---------

Co-authored-by: Scott <seascott@deloitte.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

kunal0137 committed:
fix: using the same parent id for multiple tool inputs (#1315)

kunal0137 committed:
feat: using tool names as enums for cot (#1316)

kunal0137 committed:
fix: using anyOf and sending proper map to python (#1317)

AnasSaleh-98 committed:
[#417] feat: added replayStep reactor

snehakumari369 committed:
feat(function): updated to only process files checked for dependencies (#1313)

* feat(function): updated to only go through files which it has checked for dependencies

* feat(function): made minor changes

* chore: formatting code

---------

Co-authored-by: Kumari <skumari40@deloitte.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

ryanweiler92 committed:
feat: openai image models compatible with semoss mssg format (#1318)

tlokeshrao committed:
model logs search (#1133)

* fix: model logs search

* fix: remove unused imports

* fix: remove print

* fix: casing

* fix: cast to clob

* varchar

* fix: qs minus lower

* fix: qs

* fix: active rooms

* fix: more in return

* we did it

* chain inner selectors instead

* Update ModelInferenceLogsUtils.java

Remove print

* feat: adding search room reactor

* fix: merge issues

* fix: expand to cover more db options

* remove extra selector

* get rid of other stuff too

* remove unused imports

* comment

---------

Co-authored-by: Van Buren <tevanburen@deloitte.com>
Co-authored-by: Rithvik Doshi <doshirithvik@gmail.com>
Co-authored-by: rithvik-doshi <81876806+rithvik-doshi@users.noreply.github.com>
Co-authored-by: kunal0137 <kunal0137@gmail.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
chore: adding codeowners file (#1319)

themaherkhalil committed:
chore: moving function logic to query util class, deprecating UpdateQueryStruct (#1321)

themaherkhalil committed:
feat: exposing methods on interface instead of casting IModelEngine to AbstractModelEngine (#1324)

shines-dev committed:
fix: adding PyUtils.pyEnabled() check to InsightUtility before attempting removeInsightGlobals() (#1322)

Co-authored-by: kunal0137 <kunal0137@gmail.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

AnasSaleh-98 committed:
[#417] fix: fix replayStep loop

RawanAbdelkhalek committed:
feat: list all recording

RawanAbdelkhalek committed:
Merge branch 'playwright-new-replay' of https://github.com/SEMOSS/Semoss into playwright-new-replay

kunal0137 committed:
fix: correcting timestamp used for token usage query (#1328)

kunal0137 committed:
update databricks jar for arrow changes (#1329)

update databricks jar for arrow changes

themaherkhalil committed:
chore: push log4j plugins into resources folder for eclipse to pickup during development (#1330)

themaherkhalil committed:
fix: handle types in python syntax (#1331)

themaherkhalil committed:
chore: ignore log4j plugin (#1332)

ryanweiler92 committed:
Handle tool_choice param in Python AI provider clients (#1323)

* feat: google gen ai tool_choice

* feat: support anthropic tool_choice

* feat: google gen ai - dont pass tool config if tools empty

* feat: bedrock tool_choice support

themaherkhalil committed:
fix: mcp generation, tools with complex types using openai, calls to google without tools (#1334)

* fix: properly creating the mcp json types and adding in output schema

* fix: accounting for complex mcp types to openai types

* fix: messages breaking when [] array of tools being passed

* fix: removing outputSchema since it seems to not be supported yet by some mcp clients

themaherkhalil committed:
feat: new create guardrail engine, consolidating repeated code to UploadUtilities (#1335)

themaherkhalil committed:
feat: audit database is now a temp engine instead of direct connection (#1336)

themaherkhalil committed:
feat: if the proxy is ever invoked, we will always log even if no input/output reactors (#1337)

themaherkhalil committed:
feat: switch from uuid to guidv7 (#1338)

RawanAbdelkhalek committed:
feat: ProbeElement Reactor

Signed-off-by: RawanAbdelkhalek <66391994+RawanAbdelkhalek@users.noreply.github.com>

AnasSaleh-98 committed:
[#417] feat: Added click, navigate, wait, scroll to replayStep reactor

Na-daT committed:
[#465] feat: Added cropped screenshot and LLM Context reactor

AnasSaleh-98 committed:
[#417] fix: fixed isLastPage + always add naviage step in separate array

pmagudapathy committed:
feat: log4j plugin to mask confidential information, pom update for log4j plugin to recompile even if cached on system (#1341)

* feat: Masking confidential information in the console logs and the rolling appender log files by using log4j2.LogEventPatternConverter

* feat: force compilation of log4j plugin, exposing additional regex pattern matches, moving values to mask into central location

---------

Co-authored-by: Magudapathy <pmagudapathy@deloitte.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
fix: remove inner avro from hadoop jar (#1345)

themaherkhalil committed:
fix: shade jar properly excludes conflicts from hadoop and athena (#1350)

themaherkhalil committed:
feat: reporting from audit logs (#1287)

* feat: adding in initial logic for reporting from audit logs

* feat: Changes made for fetching Timeline audit log data

* feat: modifying sql interpreter to push to query util for syntax generation of functions, changing name of AuditLog to AuditLogReport

* chore: clearer else if logic on getting the message data if Object vs Map message

* fix: query logic handles null/empty and does all filters as additions, returning all logs even if response/request is empty

---------

Co-authored-by: Magudapathy <pmagudapathy@deloitte.com>

Na-daT committed:
[#465] feat: Return probe in action list in ReplayStepReactor

mavictor24 committed:
fix(playwright): replace the text value when deleting instead of appending

resmas-tx committed:
Add step to install playwright dependencies (#1355)

Co-authored-by: kunal0137 <kunal0137@gmail.com>

AnasSaleh-98 committed:
[#417] fix: fixed executeAll logic

Na-daT committed:
[#465] Return probe data in action list for Type steps

themaherkhalil committed:
fix: better handling of classpath to compile java (#1358)

* fix: better handling of classpath to compile java

* fix: using StandardCharsets instead of string utf-8

Na-daT committed:
fix(playwright): int check for strings in probeReactor

themaherkhalil committed:
fix: removing tools from param map if empty (#1361)

themaherkhalil committed:
fix: replace old smss_driver.py file name to new mcp_driver.py file name (#1364)

* fix: replace old smss_driver.py file name to new mcp_driver.py file name

* fix: wrapping concat strings in quotes for proper replacement

themaherkhalil committed:
chore: switch from string utf-8 to StandardCharsets.UTF_8 (#1365)

kunal0137 committed:
Update anthropic_text_client.py (#1367)

themaherkhalil committed:
feat: improved streaming from python models to tomcat (#1366)

* feat: exposing new smss_stream_func in handle python to send back different operation type for streaming chunks from python models
to tomcat. consolidating methods into MCPUtility for processing tools

* chore: fixing typo in comment

* fix: making the PixelJobManager streaming thread safe

themaherkhalil committed:
chore: switch streaming logs to debug (#1368)

themaherkhalil committed:
fix: aggregating tools in semoss format, handling empty arguments (#1369)

themaherkhalil committed:
chore: removing import that is not used (#1370)

themaherkhalil committed:
chore: adding commented out method to set smss_stream when running code outside of gaas_tcp_server_handler (#1371)

themaherkhalil committed:
fix: resolving error with importing logger which was removed (#1372)

kunal0137 committed:
fix-mcp-utility-functionName (#1374)

themaherkhalil committed:
fix: adding pom.properties into shaded meta-inf resolving version error on containers (#1378)

rithvik-doshi committed:
Retrieve existing message feedback (#1118)

* refactoring message feedback init

* fix illegal start of exp

* simplify query

* pass feedback when getting messages

* extract method

* changing back to messageid and shifting loading feedback to setMessages

* getting elsa feedback almost works

* refactor

* clean up prints

* to string

* refactoring and standard return

* only add feedback if there's something to add

* moving to message utils

* Update BaseFeedback.java

* tried to clean up a bit

* renamed

* unused imports

* unused import

* feat: refactor slightly to avoid breaking error

* fix(refactor): remove ifeedback

* refactor: rename to MessageFeedback

* move parseMessage calls to account for cached vals or unnecessary calls

* only parse again after migration

* remove unnecessary parseMessages

* fix: failing when messageIds null or empty

* fix: update message json

* fix: start flushing messages to room table, add feedback directly to room in room hash + more fixes to feedback

* refactor: unused import

* switch to semoss date

* remove unused deps

* Update Room.java

Remove comment

---------

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>
Co-authored-by: kunal0137 <kunal0137@gmail.com>
Co-authored-by: Tejas Lokeshrao <tlokeshrao@deloitte.com>

tlokeshrao committed:
Workspace user perm (#1356)

* fix: workspace permissions when shared

* fix: workspace sharing

* fix: move duplicated monolith endpoint call to Security helper

* fix: workspace ownership

* fix: permission not permissions

* fix: logic error

* fix: move existing dependencies over when shared

* fix: comment for helper

* fix: remove unused method

* fix: unused import

* fix: remove unused method param

* fix: remove param from javadoc comment

* fix: authorization check prior to dependency deletion

* fix: editors on project can remove dependencies since they can add them

---------

Co-authored-by: kunal0137 <kunal0137@gmail.com>

kunal0137 committed:
Adjust COT plan id and limit tool choice options (#1382)

* fix: BE created plan id;  added COT ornament

* fix: removing some tool choice options

* feat: add step name to schema and correct required fields

---------

Co-authored-by: Scott <seascott@deloitte.com>

kunal0137 committed:
fix: using the new utils (#1383)

kunal0137 committed:
feat: adding confirm cot reactor (#1384)

pasupathimuniyappan committed:
[1342] - feat - Tool Choice for OpenAI Chat-Completion and Responses client (#1351)

* [1342] - feat - Added Tool Choice for OpenAI Chat-Completion and Responses client.

* [1342] - feat - Updated the code.

---------

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

themaherkhalil committed:
feat: exposing api_endpoint parameter (#1385)

EJV20 committed:
Unit testing updates (#1303)

* test: fixed unit tests

* test: fixed issue found while fixing unit tests

* test: fixing tests after merging latest dev

* test: removing mockito inline resource

* test: mvn test now runs unit tests with coverage

* chore: fix whitespace showing entire file as diff

---------

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
chore: removing legacy tests (#1389)

themaherkhalil committed:
feat: proper streaming (content and tools) with new OpenAI client (#1391)

* feat: extending streaming for chat completions on openai_client_v2, passing cpw into NativePySocketClient to call shutdown on disconnect to release port lock

* chore: using correct gson import

EJV20 committed:
chore: unit test fixes (#1390)

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

kunal0137 committed:
fix: passing in correct tool_choice args across models (#1393)

neelneelneel committed:
feat: allowing tools to be added to AskCOT

---------

Co-authored-by: Kunal Patel <kunal0137@gmail.com>
Co-authored-by: Scott <seascott@deloitte.com>

CodiumAI-Agent commented:
## Title

[#417] Playwright Recorder + Replayer

___
### **User description**
## Description
This feature is to allow users to record and replay browser workflows directly within Semoss. The workflow is initiated by recording a flow, generating a json file with all the steps done, and retrieving those files for replay whenever the user needs to integrate with any external system. These reactors are utilized and getting called by a react app that is going to be exposed as an MCP tool in playground.

## Changes Made
<!-- List the key changes made in this PR -->
1. Added a new package called **perna.reactor.playwright** that contains all the reactors and files needed.
2. Added a utility method **getMap** in perna.util.Utility.java as it is used by multiple reactors in playwright (and duplicated in other reactors as well)

## How to Test

1. Call Session ( ) to open a new session and use the sessionId returned in future pixel calls.
2. Call Step reactor with the session id returned to test different actions user can make (click, navigate, type, etc.)
3. Call SaveAll reactor to save the steps. (Note that currently the path is relative to project app folder, and so when testing this from terminal it will throw an error since the project name is null. **For testing purposes, you can change the folder path to any local folder to see the json file by editing the path in this reactor :ReplayFromFileReactor: initRecordingsDir**)


___

### **PR Type**
Enhancement


___

### **Description**
- Introduce Playwright recording/replay reactors

- Add DOM probe and screenshot utilities

- Support metadata listing and patching

- Extract/update script variables in JSON


___

### Diagram Walkthrough


```mermaid
flowchart LR
  Session["Session management"]
  StepExec["Step execution & history"]
  Replay["Replay from file/step"]
  Screenshot["Screenshot (crop/normal)"]
  Probe["DOM element probe"]
  Meta["Recording metadata patch/list"]
  Vars["Script variables get/update"]

  Session -- "create/get" --> StepExec
  StepExec -- "saves StepsEnvelope" --> Replay
  Replay -- "loads/writes JSON" --> Meta
  Replay -- "screenshot after replay" --> Screenshot
  Probe -- "element details" --> Replay
  Screenshot -- "cropped/normal" --> ImageCtx["Vision context"]
  Vars -- "read/write recordings JSON" --> Replay
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>30 files</summary><table>
<tr>
  <td><strong>Coords.java</strong><dd><code>Add coordinate record for actions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-0ff0e15552e9acb6d7c9c09c449be597c13289c2739cfa143d0c3b43fe1ab7b8">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ElementMetrics.java</strong><dd><code>Add element layout metrics record</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-65d0847a2b035d1aba6c01a7f08db74a492a790b64f70ba5769ccdcb63cdb7ed">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ElementProbeResponse.java</strong><dd><code>Define rich probe response model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-e7ce27f6e8ec74a6f0616e2da0660dc4ab683a97f33df15319c33755770528c6">+34/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ElementRect.java</strong><dd><code>Add element rectangle record</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-6653c2146250ed851f106f6a03e8abd983b6bfba68bd10f7018a0753da1b2ca5">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>GetPlaywrightScriptVariablesReactor.java</strong><dd><code>Reactor to extract TYPE/VARIABLE values</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-a438540376c8606e36f8f2a59a22ef55f471d34a90eabc3d3be4c604e60785cb">+129/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>ImageContextReactor.java</strong><dd><code>Vision model context for cropped images</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-297ac344a6bfa9ede7ee24df8c504758987d80f4d7d5899dc41395852fa86c1e">+111/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>ListPlaywrightScriptsReactor.java</strong><dd><code>List JSON recordings in folder</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-2707119f6243d39e35fc01d24b52312c3e0052b0ab3dc37458871cfabae390d9">+48/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>MetaPatch.java</strong><dd><code>Record for metadata patch payload</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-9ba34b7e7e8c5759d2915fa730d8e9f884be9a9d666d970be09ad56a37154fbf">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>MetadataReactor.java</strong><dd><code>Expose user metadata as map</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-8abf0a931a15ac7789e622731b4d67468a467fcd906b3f8879cec1ca66e4051b">+25/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>PatchFileMetaReactor.java</strong><dd><code>Patch metadata inside recording file</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-489ef3b6a7e60d557cb5e6134b89cc47e2da80a7f4f4702f22bd14b6061165ae">+87/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>PatchSessionMetaReactor.java</strong><dd><code>Patch in-memory session metadata</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-ec11d3c36336c7d0c1d72b12a4f890109f4175e810c2387f0c4ad3b21b9896a7">+52/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ProbeElementReactor.java</strong><dd><code>Probe DOM element at coordinates</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-2f7272d2cf731bf1f62d28bab59928e64d3cd157058f51244e1ec1fc227521c0">+234/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>RecordingMeta.java</strong><dd><code>Recording metadata record added</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-ccd32be0f59ddbd9d00a221ae7b7aeb4ddb363b586ddf89a0b77097d83586a6b">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ReplayFromFileReactor.java</strong><dd><code>Replay full recording from JSON</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-08ec036c0d23c11b966c929ba958ec1f301754fcc9c9ce9dae1f7662b04164ee">+140/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>ReplayStepReactor.java</strong><dd><code>Stepwise replay with inputs support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-2b9a120464a5e3b4527b7a9e09d5189c30d0f64f0812b959a9a14052629faf00">+205/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>SaveAllReactor.java</strong><dd><code>Save session steps with metadata</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-c0a7aa1ec711f81209b4e5ffd58b78c183eab9f15fda9aa09284f79882839648">+111/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>ScreenshotReactor.java</strong><dd><code>Take normal or cropped screenshots</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-7d093eeedf5afbd1173be57abf487584d7903754248b72bfcce86f5cda9cf8c0">+112/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>ScreenshotResponse.java</strong><dd><code>Screenshot response record added</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-4258f9a4c67cc0bb5f79bbe6fdbafe112d164c1f92d0f27cfc8bddad73b5d9fd">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Session.java</strong><dd><code>Session state and history container</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-2c8614ef7f71ba6b7e3b67101cb8e34b3c51679d0f25006eaae30644b0cd6a5b">+30/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>SessionReactor.java</strong><dd><code>Create and manage Playwright sessions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-32a939c9250df95a794b9bd1c73aed565e51c0d12f120d19af358cdfe76bbfd5">+78/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>SkipStepReactor.java</strong><dd><code>Skip current step during replay</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-cba815cc05d222c7fd0a6d9136e5b003cb1015e77838e74534a648e5df6e6961">+95/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Step.java</strong><dd><code>Define action step structure</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-3c92e2eb6b1b9642a1a49a63eeefcd2eb61facedd5b110f6fb665a5c1c4b75da">+23/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>StepReactor.java</strong><dd><code>Execute steps and track changes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-da21efe40f3cc2998c9625933e3f3725c76c77e0bc7dda5cca8484145066eb45">+218/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>StepType.java</strong><dd><code>Enumerate step action types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-6248b569a73cf31996819354b6056d916764eba6665bd87f897b4d07c5e85fec">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>StepsEnvelope.java</strong><dd><code>Envelope for steps and metadata</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-fedff12f4278c0f88892d04e72f3e7eeaba0beb1f8ee5dcf88e5da51dad3b241">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>UpdatePlaywrightScriptVariablesReactor.java</strong><dd><code>Update script variables and save</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-85e3ad1005df729523422123aafb9ba5566291fa63c9d6e0dfc1cf19d14513d1">+161/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>UserMetadata.java</strong><dd><code>User metadata record</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-2ff05875d6fe3220f57ccfc35227e6c7c3fb73c383d2b0073262bb2d68e2087f">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>VariableRecord.java</strong><dd><code>Variable record with isPassword</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-cba80739185b016b432de3dc494948ab4cd5984aafbb06205f4a2fbe4959e13a">+12/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Viewport.java</strong><dd><code>Viewport info record</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-e3eafe2936f363c0f6c975934d07f66b56d6a81118be866f260039b3e6218c3d">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Utility.java</strong><dd><code>Add helper to extract map params</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SEMOSS/Semoss/pull/1396/files#diff-baa3c1b007e2168b58afab6a8cf28a288a718bafb5441f4331f9e19f110a6b0d">+20/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

CodiumAI-Agent commented:
## PR Code Suggestions ✨

<!-- eed4f26 -->

<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Null-guard paramValues access</summary>

___


**<code>paramValues</code> can be null if no map is provided, leading to NPE when accessing <br><code>userPrompt</code> or crop params. Guard against null and missing keys, and return a clear <br>error or default. This prevents runtime crashes when optional params are omitted.**

[src/prerna/reactor/playwright/ImageContextReactor.java [43-46]](https://github.com/SEMOSS/Semoss/pull/1396/files#diff-297ac344a6bfa9ede7ee24df8c504758987d80f4d7d5899dc41395852fa86c1eR43-R46)

```diff
 Map<String, Object> paramValues = Utility.getMap(this.store, this.curRow);
+if (paramValues == null) {
+    Map<String, Object> err = new HashMap<>();
+    err.put("response", "Error: missing paramValues map");
+    return new NounMetadata(err, PixelDataType.MAP);
+}
 
-String userPrompt = (String) paramValues.get("userPrompt");
+String userPrompt = (String) paramValues.getOrDefault("userPrompt", "");
 
 Map<String, Object> result = analyzeCroppedImageWithVision(sessionId, engineId, paramValues, userPrompt);
```
<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: `Utility.getMap` can return null; accessing `paramValues.get` without checks risks NPE. Adding a guard and default improves robustness in normal flows with optional params.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Initialize insight at runtime</summary>

___


**<code>insightObj</code> is captured in the constructor before <code>this.insight</code> is initialized by the <br>framework, risking a null reference when resolving the recordings directory. Set <br><code>insightObj</code> inside <code>execute()</code> after <code>organizeKeys()</code> or pass it directly to helpers. <br>This prevents NPEs when calling <code>initRecordingsDir(insightObj)</code> or <code>loadStepsFromFile</code>.**

[src/prerna/reactor/playwright/ReplayFromFileReactor.java [41-48]](https://github.com/SEMOSS/Semoss/pull/1396/files#diff-08ec036c0d23c11b966c929ba958ec1f301754fcc9c9ce9dae1f7662b04164eeR41-R48)

```diff
 public ReplayFromFileReactor(){
 	this.keysToGet = new String[] {
 			"sessionId",
 			ReactorKeysEnum.PARAM_VALUES_MAP.getKey()
 			};
 	this.keyRequired = new int[] { 1,1 };
-	insightObj = this.insight;
 }
 
+@Override
+public NounMetadata execute() {
+	organizeKeys();
+	// ensure insight is available at runtime
+	insightObj = this.insight;
+	Map<String, Object> paramValues = Utility.getMap(this.store, this.curRow);
+
+    return new NounMetadata(replayFromFile(paramValues.get("name").toString()), PixelDataType.MAP);
+}
+
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: Capturing `insight` in the constructor can indeed be premature; moving `insightObj = this.insight` into `execute()` avoids potential null usage when initializing recordings. This is a context-correct, moderate-impact stability fix.


</details></details></td><td align=center>Medium

</td></tr><tr><td rowspan=1>General</td>
<td>



<details><summary>Harden coordinate parsing</summary>

___


**The current parsing truncates decimals and may throw <code>NumberFormatException</code> for <br>whitespace or non-numeric input. Add robust validation, trim, and safe parsing with <br>clear errors. This avoids crashes from malformed <code>coords</code>.**

[src/prerna/reactor/playwright/ProbeElementReactor.java [46-54]](https://github.com/SEMOSS/Semoss/pull/1396/files#diff-2f7272d2cf731bf1f62d28bab59928e64d3cd157058f51244e1ec1fc227521c0R46-R54)

```diff
 String[] parts = coordsStr.split(",");
 if (parts.length != 2) {
     throw new IllegalArgumentException("Invalid coords format. Expected 'x,y' but got: " + coordsStr);
 }
-//make sure the parts are integers
-parts[0] = parts[0].split("\\.")[0];
-parts[1] = parts[1].split("\\.")[0];
-return new Coords(Integer.parseInt(parts[0].trim()), Integer.parseInt(parts[1].trim()));
+String xs = parts[0].trim();
+String ys = parts[1].trim();
+try {
+    int xi = (int) Math.round(Double.parseDouble(xs));
+    int yi = (int) Math.round(Double.parseDouble(ys));
+    return new Coords(xi, yi);
+} catch (NumberFormatException nfe) {
+    throw new IllegalArgumentException("Coordinates must be numeric. Got: '" + xs + "', '" + ys + "'", nfe);
+}
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: Current parsing truncates decimals and may throw on malformed input; safer parsing with trim and numeric validation reduces runtime errors while preserving behavior.

</details></details></td><td align=center>Medium

</td></tr></tr></tbody></table>

ggaylor1 committed:
test(unit): unit tests for prerna.algorithm.learning packages (#659)

* test(unit): unit tests for prerna.algorithm.learning packages

* Merge branch 'dev' into 569-unit-test-for-prernaalgorithmlearning

---------

Co-authored-by: Adolfo Perera <adoperera@deloitte.com>

EJV20 committed:
test: added unit tests for engine.impl.owl (#1399)

EJV20 committed:
test: added tests for auth utils (#1400)

adolfo-perera committed:
test: fixed tests broken on mac (#1401)

Co-authored-by: Adolfo Perera <adoperera@deloitte.com>
Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

roxyramos committed:
575 cache unit test (#1405)

* pushing tinker test progress

* test(unit): testing against real files in memory

* tinkerUtilitiesTests needs review

* iCache tests

* file rename

* adding test setup for IGraphUtilitiesUnitTests

* adding iGraphUtilitiesUnitTests

* adding tinkerEngineUnitTests

* adding code coverage tests for TinkerEngine

* adding test coverage for JanusEngine

* adding error message to commented out  test code

* adding unit tests for CachePropFileFrameObject

* update tinkerEngine Unit tests to use tempDir

* updating tempDir for Janus tests

* adding code coverage for Icache class

* adding icache unit tests

* test(engine): fixing tests broken from dev merge

* test(cacheUtils): fixing testDeleteCache with new changes

* test(cache): testing caching insight and creating zip files

* test(cache): adding unit test for readInsightCache

* chore(tinkerUtils): removing unused imports

* test(tinker): updating tinkerEngine tests with temp folder setup

* test: removing flakey test

* test(tinker): reverting src change for filepaths in unit tests

---------

Co-authored-by: Jeff Vitunac <jvitunac@gmail.com>
Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

themaherkhalil committed:
fix: checking if room is already closed to stop accessing that chat history, switching to cache builder so chat histories clear over time (#1406)

themaherkhalil committed:
chore: Version works for local development, remove Hashtable/Vector for ConcurrentHashMap and collections synchronizedList  (#1407)

* fix: get version map info when running in local development

* chore: remove hashtable/vector for concurrenthashmap and collections synchronizedlist

* chore: removing invalid comments after refactoring

* fix: resolving codium suggestions

themaherkhalil committed:
feat: allowing anthropic tool streaming (#1408)

tlokeshrao committed:
Extra workspace features (#1388)

* fix: make sharing a one way change and only use workspace owner for permissions when workspace is not shared

* fix: get workspaces gets number of users and permission

* fix: num colabs and permission on list workspaces

* fix: add default values for private workspaces

* fix: permissions

* fix: remove comments

* fix: switch params and add project types but don't actually pass it

* fix: allow users to remove themselves from engine and project permissions table (#1404)

* fix: cleaning up projectID vs projectType filter

---------

Co-authored-by: kunal0137 <kunal0137@gmail.com>
Co-authored-by: Van Buren <tevanburen@deloitte.com>

themaherkhalil committed:
fix: tool json cleanup and additional _meta added (#1434)

* fix: tool json cleanup and additional _meta added

* fix: using details map within toolstep json

tlokeshrao committed:
fix: helper to run python methods (#1375)

* fix: workspace permissions when shared

* fix: workspace sharing

* fix: move duplicated monolith endpoint call to Security helper

* fix: workspace ownership

* fix: permission not permissions

* fix: logic error

* fix: helper to run python methods

* fix: saving progress

* fix: move existing dependencies over when shared

* fix: comment for helper

* fix: run py modules

* fix: unneeded imports

* fix: remove unused method

* fix: remove unused method

* fix: unused import

* Update SecurityProjectUtils.java

no

* Update SecurityProjectUtils.java

* Update SecurityProjectUtils.java

* fix: space instead of project

---------

Co-authored-by: kunal0137 <kunal0137@gmail.com>
Co-authored-by: rithvik-doshi <81876806+rithvik-doshi@users.noreply.github.com>

tlokeshrao committed:
Py reactor helper (#1437)

* fix: workspace permissions when shared

* fix: workspace sharing

* fix: move duplicated monolith endpoint call to Security helper

* fix: workspace ownership

* fix: permission not permissions

* fix: logic error

* fix: helper to run python methods

* fix: saving progress

* fix: move existing dependencies over when shared

* fix: comment for helper

* fix: run py modules

* fix: unneeded imports

* fix: remove unused method

* fix: remove unused method

* fix: unused import

* Update SecurityProjectUtils.java

no

* Update SecurityProjectUtils.java

* Update SecurityProjectUtils.java

* fix: space instead of project

* fix: unique module name

* fix: unused import

* fix: codium suggestion

* fix: codium 2

---------

Co-authored-by: kunal0137 <kunal0137@gmail.com>
Co-authored-by: rithvik-doshi <81876806+rithvik-doshi@users.noreply.github.com>

themaherkhalil committed:
fix: tool response coming as dict/map instead of string (#1442)

* fix: adding json loads to send / keep semoss message history with arguments as a map

* fix: sending arguments as dict, handling arguments in AskToolEngineResponse if map or string

rithvik-doshi committed:
fix: validate user login when using basic auth (#1439)

Co-authored-by: tlokeshrao <tlokeshrao@deloitte.com>

EJV20 committed:
test: UserRegistrationEmailService Unit tests (#1443)

roxyramos committed:
580 ds export tests (#1436)

* chore(code): removing unused imports

* test(exportGraph): adding unit tests for ds.export.graph

* refactor(ds.export): using connectionUtils to close connections

* test(ds.export): adding tests for rdbms and tinker graph exports

* test(ds.export): adding test for hasNextEdge

---------

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

ericgonzal8 committed:
592 unit test for prernaengine.imp.lfunction (#1438)

* test: adding unit tests for engine.impl.function package

* test: updated function tests after semoss update. Removed SentimentFunctionEngineUnitTests and  AWSTextractFunctionEngineUnitTests.

* test: cleaning up unsused imports; completing tests for 'StreamRESTFunctionEngineUnitTests'

---------

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

themaherkhalil committed:
fix: not always adding tool_choice (#1445)

* fix: not always adding tool_choice

* fix: trimming input string once

roxyramos committed:
1413 sec pass reset util (#1446)

* test(securityPasswordResetUtils): adding tests for securityPasswordResetUtils

* test(secPasswordReset): adding coverage tests for error messages

* test(SecurityUtils): adding test util to setup temp folder with security database setup and tear down.

---------

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

themaherkhalil committed:
feat: exposing multiple reactor utility methods, cleanup on nounstore method names (#1451)

* chore: adding multiple utility methods to make it easier to pull values from the nounstore

* chore: renaming methods in NounStore to be more clear and fixing references in other methods

* fix: cleaning up javadoc

roxyramos committed:
571 security token utils (#1448)

* refactor(SecurityTokenUtils): using connectionUtils to close the connection

* test(SecurityTokenUtils): adding unit tests for SecurityTokenUtilsUnitTests

---------

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
fix: adding try exception for json.loads incase tools stop loading and json is not complete (#1452)

themaherkhalil committed:
chore: switching to utility methods (#1453)

Subhadeepghosh1 committed:
fix: code changes to prevent editing teams to have same type and id (#1308)

* fix: code changes to prevent editing teams to have same type and id

* fix: removed unused method and changed logic to have same group type

* fix: changes made to allow only description edit

* fix: use equals() for groupType comparison in deleteGroupAndPropagate

* chore: improved syntax for closing wrapper iterator

---------

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

Subhadeepghosh1 committed:
feat: google calendar connectors (#1073)

---------

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

ericgonzal8 committed:
test: added SecurityUserUtils unit test class (#1449)

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
chore: adding deprecated method for some custom reactors (#1456)

themaherkhalil committed:
chore: adding deprecated method (#1463)

themaherkhalil committed:
feat: logic to predict tools outside of cot message branch; fixes for openai message building (#1464)

kunal0137 committed:
feat: updating llm to use new message builders (#1354)

* feat: updating llm to use new message builders

* chore: deprecating model rest engines; feat: converting full prompt for message jsons

* fix: writing result of full prompt to db

* chore: added support for the new parameters in python.

fix: normalized the LLM response as string format when we pass tools as empty.

* feat: updating python bindings

* feat: update python usage

* feat: langchain updates

* feat: fix command param for py bindings

* feat: handle full prompt when no command or question

* fix: converting tools in full prompt from openai

* fix: adding messages as message json

---------

Co-authored-by: pasupathimuniyappan <pasupathi.muniyappan@kanini.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>
Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

Na-daT committed:
feat: exposing single reactor to execute query against a database (#959)

* [#371] build: created generic sql query reactor

* [#371] fix: defaulted commit to false and added validation checks in SqlQueryReactor

* [#371] fix: add task optimization for select queries in SqlQueryReactor

* feat: get database category reactor (#371)

- limit sql query fetch to 50 rows
- create getDatabaseCategory to identify sql or noSQL db

* fix: inner reactors have their own independent noun stores, consolidating some methods

* chore: reverting back reactor factory changes

---------

Co-authored-by: Ibrahim El Nemr <70034333+ibrahimelnemr@users.noreply.github.com>
Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
feat: new reactor to loop through every page of a pdf and convert to an image (#1467)

themaherkhalil committed:
fix: only returning the file names created (#1468)

kunal0137 committed:
fix: using this_content_block instead of tool_call (#1470)

tlokeshrao committed:
modern way of get room options (#1469)

* fix: modern way of get room options

* fix: handle null

* fix: fix test

rithvik-doshi committed:
Fix load module (#1462)

* fix: load module looks at right folder

* chore: adding execution insight to track which insight/user ran the command, args as Object to handle proper translation to python object syntax

* chore: adding py in appFolder init

---------

Co-authored-by: Maher Khalil <themaherkhalil@gmail.com>

themaherkhalil committed:
feat: bedrock tool streaming, proper closing of process when socket terminates  (#1471)

* fix: properly shutting off and releasing the process

* feat: exposing streaming with tools on bedrock

ericgonzal8 committed:
test: added SecurityShareSessionUtils unit tests; fixed bug in AbstractSecurityUtilsUnitTests (#1466)

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

EJV20 committed:
1412 security api user utils unit tests (#1472)

* test: created test class

* test: fixed tests, and added SecurityApiUserUtils unit tests

---------

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

EJV20 committed:
test: update to junit 6 (#1474)

Co-authored-by: Ryan Weiler <ryanweiler92@gmail.com>

themaherkhalil committed:
fix: proper validation checks in AbstractEngineFileReactor (#1475)

* fix: proper check if user is not publisher

* fix: admin only functionality check on creation of python function engine

Na-daT committed:
feat: Add playwright utils

Na-daT committed:
feat: HTML Extraction & LLM Step Generation

Na-daT committed:
Merge branch 'dev' of https://github.com/SEMOSS/Semoss into recovery-branch

kunal0137 committed:
fix: pushing compile fixes

AnasSaleh-98 committed:
feat: Added extracting selector logic

kunal0137 committed:
Merge branch 'dev' into playwright-new-replay

AnasSaleh-98 committed:
feat: Added skip step for replay

ibrahimmahaba committed:
Merge branch 'dev' into playwright-new-replay

Na-daT committed:
fix: Prompt and HTML Extraction enhancements

RawanAbdelkhalek committed:
fix: session utility dynamic wait time update

kunal0137 committed:
Merge branch 'dev' into playwright-new-replay

kunal0137 commented on src/prerna/reactor/playwright/SessionReactor.java:
```diff
@@ -0,0 +1,71 @@
+	
+	private static final Logger classLogger = LogManager.getLogger(SessionReactor.class);
+	private Browser browser;
+    public final static Map<String, Session> sessions = new ConcurrentHashMap<>();
```
Can this piece be moved into the user object to get the session? This can't be in the reactor. if not in the user then it needs to go into a utility method.

kunal0137 committed:
moving session to user object

Na-daT committed:
Merge branch 'dev' into playwright-new-replay

AnasSaleh-98 committed:
fix: remove outdated reactors

AnasSaleh-98 committed:
Merge branch 'dev' into playwright-new-replay
